docs/internal/abi.md

```diff
@@ -0,0 +1,444 @@
+Matching against a concrete type is a simple comparison: does the descriptor pointer in the object point to the descriptor for the concrete type?
+Matching a pointer to an unknown concrete type (i.e. whose static type is an interface type) against an interface type requires a more complex lookup.
+
+The remaining fields can, on internal representations, be packed or reordered for optimisation purposes.
+Machine-word types (ex. numeric) are always represented by value (their singleton representation).
+
+For example, a class:
+```ts
+class Other { ... }
+class Foo
+{
+    var a : U32 & imm;
+    embed var b : F64 & imm;
+    var other : Other & mut;
+    create() { ... }
+}
+```
+
+A naive layout could be:
+```ts
+ { { ptr, ptr }, i32, f64,  ptr   }
+      header,    U32, F64, Other*
+```
+
+With `ptr` 32/64/128-bits, depending on the architecture.
+
+As a future optimisation, we could reorder the fields by size.
+Some targets have stronger alignment requirements, and unaligned reads can incur in performance penalties.
+Having an 8-bit type between 64-bit types can misalign the larger objects.
+By sorting the types by size, we guarantee that all 64-bit values are 64-bit aligned, all 32-bit values are 32-bit aligned and so on.
+
+A more optimal layout for the same class above, on either 32 or 64 bit targets, would be:
+```ts
+ { { ptr, ptr }, f64,   ptr,  i32 }
+      header,    F64, Other*, U32
+```
+
+Because `f64` aligns stronger than `i32` and `ptr` can be either.
+On CHERI, the 128-bit `ptr` would be the first field after the header.
+
+### Dispatch Tables & Selector Colouring
+
+Object headers have a pointer to their type's dispatch table.
+A dispatch table contains pointers to the functions that the type provides, at specific offsets.
+Those pointers are called when it's not possible to determine the actual function being called at compile time.
+There is only one dispatch table per type, not per object.
+
+The offset of each method is calculated at compile time.
+The address at each offset has a pointer to the actual functions.
+Dynamic dispatch is done by taking the pointer at the offset from a table pointer and calling that.
+If an interface provides a method `foo`, all classes that implement that interface will have a method `foo` and their dispatch tables will have an entry for it.
+
+To avoid each type to have a different calculation for each method's offset, the compiler will do **selector colouring**.
+This process finds all common methods across all types and ensure the same methods all have the same offsets on all tables.
+This applies to concrete classes that implement specific interfaces, but it's not limited by it.
+Any two classes that have the same methods (signature) will end up with the same offset.
+
+With multiple interfaces and classes implementing similar methods, this can lead to a number of gaps between offsets.
+The tables will be compacted (either during colouring or afterwards) to minimise that.
```

Comment:
By aliasing disjoint sets of selectors?

Reply:
It's unclear to me all methods by which we can compact the table, and which of them have to be part of the original colouring algorithm or done after the colouring is finished. My initial pessimistic assumption is that any meaningful compacting will happen during colouring, which makes it much more complex.

Reply:
Right, this is what I meant.

Colouring being done early makes it harder to compact later because trying to find matching sets in a sea of overlapping sets is a practically unbound problem. Not to mention having to change the indexes on flattened IR code after-the-fact.

So, compacting is done by colouring, but how far do we go during colouring or in which graph we run it is a challenge.